Update readme and add code coverage test

cpjordan opened the pull request:
This pull request updates the main README and adds a README to the example repo to provide short descriptions for each example in the examples/ directory.

Additionally, this PR adds code coverage.

These changes aim to improve the introduction of the project for new users and make it feel a bit more welcoming.

stephankramer commented on .github/workflows/build.yml:
```diff
@@ -57,24 +57,47 @@ jobs:
+        uses: codecov/codecov-action@v3
+        with:
+          files: ./coverage.xml
+          dry_run: true  # fail_ci_if_error: true
```
What does this do?

stephankramer reviewed (approved):
Looks good, nice! It looks like the turbines, optimisation and inversion modules could do with some more coverage. What does the codecov upload dry run stuff do? I've just added the codecov app to the repo, but it doesn't seem to have changed anything under https://app.codecov.io/github/thetisproject/thetis

Not to be a stickler for rules, but the README changes would have made more sense as a separate PR. All looks good though

cpjordan commented:
> Looks good, nice! It looks like the turbines, optimisation and inversion modules could do with some more coverage. What does the codecov upload dry run stuff do? I've just added the codecov app to the repo, but it doesn't seem to have changed anything under https://app.codecov.io/github/thetisproject/thetis
> 
> Not to be a stickler for rules, but the README changes would have made more sense as a separate PR. All looks good though

Yes sorry about that - I probably should have created an Issue first as well. I just wanted to generate something so that we could discuss it the next time we have a Thetis meeting as I thought we could have something nicer to encourage more first time users. I'll split it into two PRs now because the code coverage has a bit more work to do anyway.

The `dry_run` flag tells the Codecov GitHub Action to simulate the upload process without actually sending the coverage report to Codecov’s servers. (see https://github.com/codecov/codecov-action for more info on other flags). It should validate that the coverage files (coverage.xml in this case) are found and can be processed, checks that the action is correctly configured (e.g., any tokens are set if needed) but does not upload any data to Codecov. I did this in case there was a reason we didn't already do code coverage. I think to actually upload it, one of the core team members might need to connect the repo (or maybe I can - I'll find out). I'll create a separate PR now.

stephankramer commented:
Great. I've read through the README's and they look very good, I think this is a nice addition for people that arrive at the repository before they've looked at the thetisproject.org page. Could I make one further suggestion: for these people it might also be helpful to have a README in the demos/ directory? At the moment it's a bit unclear from just looking at the directory structure what the difference between demos/ and examples/ is, where the former are some basic "demo tutorials" that are written in sphinx and can be viewed directly from the website. So maybe just one or two lines saying that with a link to https://thetisproject.org/documentation.html

Oh and there's two notebooks as well, which apparently are not mentioned on the website. Maybe we should add two colab links on https://thetisproject.org/documentation.html to run these on colab, just like on https://www.firedrakeproject.org/notebooks.html. Let me just issue that....

cpjordan commented:
> Could I make one further suggestion: for these people it might also be helpful to have a README in the demos/ directory?

Generated a similar readme for the demos folder - did similar detail to the examples readme for those that skipped the website to run things locally.

> Maybe we should add two colab links on https://thetisproject.org/documentation.html to run these on colab, just like on https://www.firedrakeproject.org/notebooks.html.

Updated the documentation and notebooks so that they can be run in colab (tested, works for me) online the same way as the Firedrake examples. Will update the website repo afterwards if everything is fine here!

stephankramer commented:
Wow that's amazing. Wasn't expecting you to actually do the colab thing in this PR already, but that's great thanks!

I've just tried out the notebook on colab using: https://colab.research.google.com/github/thetisproject/thetis/blob/update_readme/demos/02-2d-tsunami.ipynb
Unfortunately it didn't work for me yet, getting:

```
"/tmp/firedrake-install.sh": No such file or directory

---------------------------------------------------------------------------

ModuleNotFoundError                       Traceback (most recent call last)

[/tmp/ipython-input-3-2447761698.py](https://localhost:8080/#) in <cell line: 0>()
      1 try:
----> 2     from firedrake.pyplot import triplot, tricontourf, quiver
      3     from thetis import *

ModuleNotFoundError: No module named 'firedrake'


During handling of the above exception, another exception occurred:

ModuleNotFoundError                       Traceback (most recent call last)

[/tmp/ipython-input-3-2447761698.py](https://localhost:8080/#) in <cell line: 0>()
      4 except ImportError:
      5     get_ipython().system('wget \\"[https://fem-on-colab.github.io/releases/firedrake-install-release-real.sh\\](https://fem-on-colab.github.io/releases/firedrake-install-release-real.sh//)" -O \\"/tmp/firedrake-install.sh\\" && bash \\"/tmp/firedrake-install.sh\\"')
----> 6     from firedrake.pyplot import triplot, tricontourf, quiver
      7     get_ipython().system('git clone https://github.com/thetisproject/thetis.git /content/thetis')
      8     get_ipython().run_line_magic('cd', '/content/thetis')

ModuleNotFoundError: No module named 'firedrake'


---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.
---------------------------------------------------------------------------
```

so it seems to have gone past the `!wget ...` line, but then despite successfully (?) running wget, it still seems to not find the script at /tmp/firedrake-install.sh ? Any idea?

stephankramer commented:
Oh looks like the `\"` quotes shouldn't be there. If I change that line to:

```
!wget https://fem-on-colab.github.io/releases/firedrake-install-release-real.sh -O /tmp/firedrake-install.sh && bash /tmp/firedrake-install.sh
```
it works for me

cpjordan commented:
> Oh looks like the `\"` quotes shouldn't be there

Ah sorry, this comes from the way I was updating/saving it to avoid run errors. Updated & checked it was running locally and broke CI due to kernel name, updated instead using a text editor (where the `\"` came from) and broke CI due to encoding, fixed that and then forgot to check it was working on Colab again!

> Ah, but now the next issue is that the runfiles/ aren't there. 

No problem - currently we install Thetis on Colab by cloning the repo so we can just change directory afterwards i.e. just add these lines:
```
    %cd /content/thetis/demos
    print(f"Firedrake and Thetis installed! The current working directory is {os.getcwd()}, but please note these files (including this one!) are all temporary.")
```

But perhaps it would be preferable to do it this way:
```

import os
try:
    from firedrake.pyplot import triplot, tricontourf, quiver
    from thetis import *
except ImportError:
    print("Firedrake or Thetis failed to import, we assume you are using Colab!")
    !wget "https://fem-on-colab.github.io/releases/firedrake-install-release-real.sh" -O "/tmp/firedrake-install.sh" && bash "/tmp/firedrake-install.sh"
    from firedrake.pyplot import triplot, tricontourf, quiver
    !pip install git+https://github.com/thetisproject/thetis.git
    from thetis import *
    demo_dir = "/content/tsunami_demo"
    runfiles_dir = os.path.join(demo_dir, "runfiles")
    os.makedirs(runfiles_dir, exist_ok=True)
    os.chdir(demo_dir)
    base_url = "https://raw.githubusercontent.com/thetisproject/thetis/master/demos/runfiles"
    filenames = ["bathy.h5", "mesh.msh", "outputs.txt"]
    for fname in filenames:
        output_path = os.path.join(runfiles_dir, fname)
        if not os.path.exists(output_path):
            print(f"Downloading {fname}...")
            !wget -q "{base_url}/{fname}" -O "{output_path}"
    print(f"Firedrake and Thetis installed! The current working directory is {os.getcwd()}, but please note these files (including this one!) are all temporary.")
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import scipy.interpolate  # used for interpolation
import pyproj  # used for coordinate transformations
import math
```

I've checked both work. Happy to do it in this PR as the demos README references these scripts. If you prefer to keep things separate I'll just do a new PR and update the tiny bit of the README that references the Colab bit there.

Let me know a) preferred Colab Thetis installation method and b) separate PRs or keep combined. Happy either way on both!

stephankramer commented on demos/02-2d-tsunami.ipynb:
```diff
@@ -36,15 +36,52 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, if we are in Colab we need to download some necessary files for this example to work. If you are not using Colab, we expect that you are in the Thetis demos repository which should already have these files."
```
```suggestion
    "Now, if we are in Colab we need to download some necessary files for this example to work. If you are not using Colab, we expect that you are in the demos/ directory of the Thetis repository which should already have these files."
```

stephankramer reviewed (dismissed):
Excellent
